### ISSUE TITLE
Change Display Name of Testcase ( Theory )

### ISSUE THREAD

**Issue by arpitgold** [xunit/xunit#649]
Hi,

I have created test case with Theory. and I am passing Custom Object to test method.
After execution of my test, I see unmeaningful test case name.

![](http://i.imgur.com/G4q0RiY.png)

**Current Testcase name** : Negative[Null FareSapce]-2 + (complete object data)

**Required Testcase name**  : 1- Negative[Null FareSpace]-2
I want to append on specific data from parameter object.

Thanks
Arpit

**Comment by bradwilson** [comment]
You can control the rendering of a data object in the theory display by overriding the `ToString()` method on the data object.

**Comment by arpitgold** [comment]
Can you please give some example?

**Comment by bradwilson** [comment]
```csharp
public class SupplierMapping
{
    public string ID { get; set; }
    //...other properties...

    public override virtual string ToString()
    {
        return string.Format("ID #{0}", ID);
    }
}
```

would show in display as:

```
Test Display Name(supplier: ID #42)
```

**Comment by jongleur1983** [comment]
This is possible, of course, but IMHO it's not sufficient for some use cases.
ToString() for a particular class might be overwritten already to show stuff that's most commonly used independent of tests.
As a test covers a particular, usually small scoped part of that class, it may be very different what's relevant for the test case than what's relevant when you usually show your object anywhere.

Unfortunately an attribute cannot have a lambda as property, but it could be an attribute linking to a method of the test class with the signature ParameterType -> string, similar to the existing MemberData attribute.

**Comment by arpitgold** [comment]
hi @bradwilson,
Thanks for your solution and sorry for my late comment
I found following issues in the approach :
1. I have multiple parameters in test case, so it output as testcase name(formatted data1, formatted data2...... and so).
I don't want to show some parameters, but I can not control it with this approach.

2. Some of the parameter objects are 3rd party classes, so I can not override ToString() for those classes. so it still show the full object values for those classes.

I want to override **displayname** and want to write test case name differently.
Is there any way so that I can inject values in Before and After test case and use method(test case) parameter and over ride display Name?

Thanks
Arpit Nagar

**Comment by nbarbettini** [comment]
I too would like a way to specify a custom displayName for a theory. It's an edge case but would be nice in some situations. For instance, I have a few theories that take a lot of parameters, or long strings as parameters.

Maybe something like:
```
[Theory]
[InlineDataWithName("My Theory", obj1, obj2)]
(...)

[Theory]
[MemberDataWithNames(nameof(MyTheoryNames), nameof(TestCases))]
(...)
```
Not sure if that syntax is too obscure, though.

Currently I've worked around this in two ways:
* Passing a dummy string containing the name as the first parameter
* Representing the entire test case as a POCO with an overridden `ToString` method

These work fine, so it's not something I need badly.

**Comment by netchkin** [comment]
Hi, I particularly like the `InlineDataWithName` idea, but would incline more to the way it is done in NUnit - via Attribute parameter, like this:

```
[Theory]
[InlineData(obj1, obj2, true, DisplayName = "working with obj1 and ob2 always yields correct result")
(...)
```

I think it's more readable than any other solution (and also implementable - the alternative would be adding a parameter to the constructor, which is currently not possible without hindering readability)

**Comment by jongleur1983** [comment]
I like the idea of @netchkin, but I am not sure if it suggests the right way to implement tests to developer.

IMHO the displayname of any given "run" of a theory should follow the same systematic schema - the same as the the theory should check the same facts in any given "run" as well. Therefore the DisplayName should not be defined on an individual theory run (like in InlineData), but in a set of runs - either the TheoryAttribute itself or the single data source like MemberDataAttribute.

On the other hand it is indeed good to have some way to include parametrization to the theories to make the individual parameters more readable.

That's why I would move the DisplayName from the DataAttribute to the Theory attribute itself:

```[Theory(DisplayName = "For MyComparer vehicle and motor vehicle are equal.")]```

with parameters in the DisplayName this would lead to

```[Theory(DisplayName = "For MyComparer the vehicle {0} and motor vehicle {1} are equal.")]
[InlineData(vehicle, motorVehicle)]
public void VehicleComparerEquals(Vehicle v, MotorVehicle mv) {...}```

where the DisplayName for the given inlineData would be generated by xUnit using 
```string.Format(attribute.DisplayName, theoryParams)``` (theoryParams is the array of parameters the Theory method is called with in each run).

This would suggest to the developer who uses XUnit that a theory is still one theory, not a set of theories. It follows the old scheme too to display a common name for all instances of the same theory and to add the parameters in the displayed "name" as well.

**Comment by D0tNet4Fun** [comment]
Hi,

I would like to have the ability to change the display name too, or at least have a way to avoid the rendering of test parameters. I use AutoFixture with NSubstitute and a test of mine may look like this:

`public void MethodToTest(  [Frozen] IDependency1 dep,   [Frozen] IDependency2 dep2,     SystemUnderTest sut)`

Xunit displays the test as:
`MethodToTest(dep: IDependency1Proxy {_interceptors = [CastleForwardingInterceptor {...} ... ` - which is not very helpful. I'd rather it reads `MethodToTest(dep, dep2, sut)` which I think is how it was in xunit1. Maybe something like this can help:

`[Theory(DisplayNameFormat=DisplayNameFormat.Legacy, ... ]`
Other values can be: 
-  Default - for the new 2.0 format
- Custom - in which case the user needs to specify the DisplayName just like @jongleur1983 suggested.

**Comment by bradwilson** [comment]
There has been a lot of discussion here, but it's not clear what people want. Please open new issues with specific feature requests.

**Comment by bradphelan** [comment]
The solution is just to add a layer of indirection.

    public class Labeled<T>
    {
        public Labeled(T data, string label)
        {
            Data = data;
            Label = label;
        }

        public T Data { get; }
        public string Label { get; }

        public override string ToString()
        {
            return Label;
        }
    }

    public static class Labeledextensions
    {
        public static Labeled<T> Labeled<T>
                (this T source, string label)=>new Labeled<T>( source, label );
    }

    
And one of my production tests

    public class ControlPointSolvedInterpolatorDistanceSpec
    {
        public static IEnumerable<object[]> Data()
        {
            object[] make
                ( ImmutableList<IControlPointSolved> items
                , double distance
                , string label)
            {
                return new object []
                { ( cpsi:new ControlPointSolvedInterpolator( items )
                  , distance:distance
                  )
                  .Labeled( label ) 
                };
            }

            ICurve curve = null;
            {
                var items = ImmutableList.Create(
                          CreatToolPathItem(x: 0, y: 0, z: 0, c: 0, b: 0, data: "0".dumb(), railPosition: 0, curve: curve),
                          CreatToolPathItem(x: 10, y: 0, z: 0, c: 0, b: 0, data: "3".dumb(), railPosition: 9, curve: curve)
                     );

                yield return  make( items, 10, "x:10" );
            }

            {
                var items = ImmutableList.Create(
                          CreatToolPathItem(x: 0, y: 0, z: 0, c: 0, b: 0, data: "0".dumb(), railPosition: 0, curve: curve),
                          CreatToolPathItem(x: 0, y: 10, z: 0, c: 0, b: 0, data: "3".dumb(), railPosition: 9, curve: curve)
                     );

                yield return  make( items, 10, "y:10" );
            }

            {
                var items = ImmutableList.Create(
                          CreatToolPathItem(x: 0, y: 0, z: 0, c: 0, b: 0, data: "0".dumb(), railPosition: 0, curve: curve),
                          CreatToolPathItem(x: 0, y: 0, z: 10, c: 0, b: 0, data: "3".dumb(), railPosition: 9, curve: curve)
                     );

                yield return  make( items, 10, "z:10" );
            }

            {
                var items = ImmutableList.Create(
                          CreatToolPathItem(x: 10, y: 0, z: 0, c: 0, b: 0, data: "0".dumb(), railPosition: 0, curve: curve),
                          CreatToolPathItem(x: 10, y: 0, z: 10, c: System.Math.PI*2, b: 0, data: "3".dumb(), railPosition: 9, curve: curve)
                     );

                yield return  make( items, 10 * 2 * Math.PI, "x:10, c:2Pi" );
            }

        }


        [Theory]
        [MemberData(nameof(Data))]
        public void ShouldCalculateDistance(Labeled<(ControlPointSolvedInterpolator cpsi, double distance)> t)
        {
            t.Data.cpsi.TotalDistance.Should().BeApproximately( t.Data.distance, 1e-5 );
        }

    }
    
    
![image](https://user-images.githubusercontent.com/17650/29511949-132b988a-8661-11e7-843e-f9366a0f3517.png)

Previously I couldn't tell the difference between any of the theories because the ToString() methods on the original data objects was too comprehensive and the signal is lost in the noise.

**Comment by ChaosCrafter** [comment]
I sometimes use a first parameter of string _for
For example...
`            [Theory]
            [InlineData("null", null, null)]
            [InlineData("empty", "", "")]
            [InlineData("lowercase", "Fred Bloggs", "fred bloggs")]
            [InlineData("uppercase", "Fred Bloggs", "FRED BLOGGS")]
            [InlineData("mixed_case", "Fred Bloggs", "fReD bloGGs")]
            public void Capitalisation_is_correct(string _for, string expectedString, string actualString)
            {
                Assert.Equal(expectedString, actualString.ToTitleCase());
            }`

This means my tests get labeled 
`Capitalisation_is_correct(_for: "uppercase', ...`
Which is at least easy to quickly read, even if it's not ideal.

I really like @jongleur1983's idea with a test DisplayName format.
My only change is I'd like to also have an option on extending the test function name as I think you should be able to easily find the test by name directly.
So valid options would be 
[Theory(DisplayName = "For MyComparer the vehicle {0} and motor vehicle {1} are equal.")] 
[InlineData("car1","car2")] 
public void VehicleComparerEquals(Vehicle v, MotorVehicle mv) 
which would display as @jongleur1983 presented
"For MyComparer the vehicle car1 and motor vehiclecar2 are equal."

and
[Theory(DisplayNameExtended = "vehicle {0} and motor vehicle {1} are equal.")] 
public void VehicleComparerEquals(Vehicle v, MotorVehicle mv) 
which would display
"VehicleComparerEquals vehicle car1 and motor vehicle car2 are equal."

**Comment by bkqc** [comment]
For MemberData, could there be a new attributes used say NamedMemerData? This way, there could be some extension method on Object[]. Something like
```
public static NamedCase AddDisplayName(this object[] data, string displayName){...}

public Class NamedCase{
  public object[] Data {get; set;}
  public string DisplayName{get; set;}
}
```
XUnit could then use the DisplayName for results and still pass Data to the method.

Another option would be that  instead of having a different attribute, the extension method could return object[] containg a NamedCase object and XUnit could then do a special treatment on it. No breaking change, all transparent and it would also allow to have, in a single data-driven test,  some custom named tests and other using the automatic naming.

**Comment by bkqc** [comment]
Another variation could be to have a different attribute, say NamedMemberDataAttribute that would expect a return type of  `IEnumerable<(string, object[]>`

**Comment by commonquail** [comment]
This issue was closed nearly two years before this comment, citing a solution that doesn't work well for `ClassDataAttribute` in a version of the library that has not been released in the meantime.

To summarize the options available as of xunit 2.4.2, ordered for reference:

1. Override the data parameter type's `ToString()`.
   * This is a very flexible and approachable solution _where possible_, and completely useless anywhere the parameter type's `ToString` is already constrained.
1. Pass a constant string for context as the first data parameter.
   * This works for any `DataAttribute` data source but dynamic labeling has to be performed by hand up front.
1. Pack the logical data parameter type inside a tuple together with a string that is formatted before test case evaluation.
   * This is the runtime evaluated version of no.2), allowing automating the labeling of data parameters.

Each solution has its drawbacks. I prefer no.3) for its generalizability and relative type safety but it requires the considerable boilerplate.

---

Using effectively bradphelan's

```csharp
public sealed record Labeled<T1, T2>(string Label, T1 Data1, T2 Data2)
{
    public override string ToString() => Label;
}
```

we can make an alternative `TheoryData<T1, T2>` implementation that eagerly labels the logical data parameter value

```csharp
public abstract class LabeledTheoryData<T1, T2> : TheoryData
{
    public void Add(T1 p1, T2 p2)
    {
        var formattable = Label(p1, p2);
        var label = formattable.ToString(CultureInfo.InvariantCulture);
        AddRow(new Labeled<T1, T2>(label, p1, p2));
    }

    protected abstract FormattableString Label(T1 p1, T2 p2);
}
```

and use it like this:

```csharp
internal record MyType(IReadOnlyList<string> Foo);
internal record MyOtherType(int Bar);

internal class MyData : LabeledTheoryData<MyType, MyOtherType>
{
    protected override FormattableString Label(MyType p1, MyOtherType p2)
    {
        return $"Foo.Count={p1.Foo.Count}), Bar={p2.Bar}";
    }

    public MyData()
    {
        Add(new MyType() { Foo = new string[] { "a", "b", }}, new MyOtherType(42));
    }
}
```

**Comment by rbeurskens** [comment]
Is there a given reason why ```ITheoryDataRow``` is not (going to be) backported to xUnit2? That would make it so much easier. I like to be able to port my NUnit tests which have a ```TestCaseData``` class covering this role.
